pkg/dns/service_test.go

```diff
@@ -360,3 +416,436 @@ func TestDNS_findMatchingManagedZone(t *testing.T) {
+				ObjectMeta: v1.ObjectMeta{
+					Namespace: "test",
+				},
+			},
+			mz: &v1alpha1.ManagedZoneList{
+				Items: []v1alpha1.ManagedZone{
+					{
+						ObjectMeta: v1.ObjectMeta{
+							Name:      "example.com",
+							Namespace: "test",
+						},
+						Spec: v1alpha1.ManagedZoneSpec{
+							DomainName: "example.com",
+						},
+					},
+				},
+			},
+			scheme:        testScheme(t),
+			wantSubdomain: "example.com",
+			wantErr:       false,
+		},
+		{
+			name: "unable to list MZ",
+			host: "example.com",
+			gateway: &gatewayv1beta1.Gateway{
+				ObjectMeta: v1.ObjectMeta{
+					Namespace: "test",
+				},
+			},
+			mz:      &v1alpha1.ManagedZoneList{},
+			scheme:  runtime.NewScheme(),
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			gw := traffic.NewGateway(tt.gateway)
+			f := fake.NewClientBuilder().WithScheme(tt.scheme).WithLists(tt.mz).Build()
+			s := dns.NewService(f, &fakeHostResolver{})
+
+			gotMZ, gotSubdomain, err := s.GetManagedZoneForHost(context.TODO(), tt.host, gw)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("GetManagedZoneForHost() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !tt.wantErr && !reflect.DeepEqual(gotMZ.ObjectMeta, tt.mz.Items[0].ObjectMeta) {
+				t.Errorf("GetManagedZoneForHost() gotMZ = %v, want %v", gotMZ, tt.mz.Items[0])
+			}
+			if gotSubdomain != tt.wantSubdomain {
+				t.Errorf("GetManagedZoneForHost() gotSubdomain = %v, want %v", gotSubdomain, tt.wantSubdomain)
+			}
+		})
+	}
+}
+
+func TestService_SetEndpoints(t *testing.T) {
+
+	tests := []struct {
+		name      string
+		addresses []gatewayv1beta1.GatewayAddress
+		dnsRecord *v1alpha1.DNSRecord
+		wantErr   bool
+	}{
+		{
+			name: "endpoints are set",
+			addresses: []gatewayv1beta1.GatewayAddress{
+				{
+					Type:  addr(gatewayv1beta1.IPAddressType),
+					Value: "1.1.1.1",
+				},
+				{
+					Type:  addr(gatewayv1beta1.HostnameAddressType),
+					Value: "example.com",
+				},
+			},
+			dnsRecord: &v1alpha1.DNSRecord{
+				ObjectMeta: v1.ObjectMeta{
+					Name: "example.com",
+				},
+			},
+		},
+		{
+			name: "no new endpoints",
+			addresses: []gatewayv1beta1.GatewayAddress{
+				{
+					Type:  addr(gatewayv1beta1.IPAddressType),
+					Value: "1.1.1.1",
+				},
+			},
+			dnsRecord: &v1alpha1.DNSRecord{
+				ObjectMeta: v1.ObjectMeta{
+					Name: "example.com",
+				},
+				Spec: v1alpha1.DNSRecordSpec{
+					Endpoints: []*v1alpha1.Endpoint{
+						{
+							DNSName:       "example.com",
+							SetIdentifier: "1.1.1.1",
+							Targets:       []string{"1.1.1.1"},
+						},
+					},
+				},
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			f := fake.NewClientBuilder().WithScheme(testScheme(t)).WithObjects(tt.dnsRecord).Build()
+			s := dns.NewService(f, &fakeHostResolver{})
+
+			if err := s.SetEndpoints(context.TODO(), tt.addresses, tt.dnsRecord, &v1alpha1.DNSPolicy{}); (err != nil) != tt.wantErr {
```

Comment:
do we not want to validate that the DNSRecord has been changed as expected?

Reply:
Good point. Done